### PR TITLE
ENGA-1457c "handle different field types in replace null with empty"

### PR DESCRIPTION
**Pull request by DevDaveFrame**
[ENGA-1457](https://industrydive.atlassian.net/browse/ENGA-1457)

### Non-technical Context
This change improves how null values are handled when exporting data to BigQuery, preventing errors that occur when empty strings are assigned to numeric fields.

### Technical Context
1. **bigquery_exporter/constants.py:** Created a new constants file that contains a dictionary mapping BigQuery field types to appropriate falsy default values (e.g., 0 for INTEGER, 0.0 for FLOAT, False for BOOLEAN, etc.).

2. **bigquery_exporter/base.py:** Enhanced the `_sanitize_value` method to handle None values based on field types. When `replace_nulls_with_empty` is enabled and a None value is encountered, the method now:
   - Checks the BigQuery schema to determine the field's type
   - Returns the appropriate type-specific default (0 for integers, 0.0 for floats, etc.) based on the field type
   - Falls back to empty string if the field isn't found or has an unknown type

3. **tests/test_base.py:** Added tests to verify the new type-specific default functionality:
   - Testing different field types get appropriate defaults (strings, integers, floats, booleans, etc.)
   - Verifying that non-None values remain unchanged
   - Ensuring fields not in the schema get handled gracefully
   - Confirming that disabling `replace_nulls_with_empty` still preserves None values

### Manual Testing Instructions
To test this functionality:
1. Create a BigQueryExporter with `replace_nulls_with_empty=True`
2. Export a model with null values in various field types
3. Verify that each null value is replaced with the appropriate default (e.g., 0 for INTEGER fields, empty string for STRING fields)

#### Developer

- [x] Ensure code complies with the [style guide](https://industrydive.atlassian.net/wiki/spaces/TECH/pages/639500407/Tech+Team+s+Style+Guides)
- [x] Devise edge cases to test the bug fix or feature being merged
- [x] Make sure the code is clean and understandable
- [x] Ensure code is not overly inefficient
- [ ] Ensure user permissions have been included (for new features) or updated (for revisions), if necessary

#### Reviewer
- [ ] Ensure code complies with the [style guide](https://industrydive.atlassian.net/wiki/spaces/TECH/pages/639500407/Tech+Team+s+Style+Guides)
    - If there are few or minor issues, add inline comments
    - If there are many issues or it seems like the developer did not follow the style guide, leave a comment asking them to do so
- [ ] Devise edge cases to test the bug fix or feature being merged
- [ ] Make sure the code is clean and understandable
    - [ ] Ask for inline comments on unclear sections
- [ ] Ensure code is not overly inefficient
- [ ] Ensure user permissions have been included (for new features) or updated (for revisions), if necessary
- [ ] Ensure documentation is understandable and accurate, including:
    - [ ] Code comments and doc strings
    - [ ] Technical documentation
- [ ] There are no conflicting migration leaf nodes

[ENGA-1457]: https://industrydive.atlassian.net/browse/ENGA-1457?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: bigquery_exporter/base.py [hunk 1]
```diff
@@ -246,8 +248,19 @@ def _sanitize_value(self, value):
...
+        elif value is None and self.replace_nulls_with_empty:
+            # If we don't have a field name, default to empty string
+            if field_name is None:
+                return ''
+
+            # Try to find the field in the table schema
+            for schema_field in self.table.schema:
+                if schema_field.name == field_name:
+                    # Return the appropriate default for this field type
+                    return BQ_TYPE_DEFAULTS.get(schema_field.field_type, '')
```
Comment (ddicocco) [inline bigquery_exporter/base.py]: Just a suggestion, (take or leave as i'm not sure how much more performant it'd be in practice): i think this could benefit from using a mapping of schema field names to default values here instead of potentially iterating through the schema every time `_sanitize_value` is called. In practice, I'd think this `for` is fast, it's just the idea that we're doing a lot of potentially redundant looping in the (not unlikely?) case where we have large schema + null values

File: bigquery_exporter/base.py [hunk 2]
```diff
@@ -246,8 +248,19 @@ def _sanitize_value(self, value):
             return handle_datetime_value(value)
         elif isinstance(value, UUID):
             return str(value)
-        elif value is None:
-            return '' if self.replace_nulls_with_empty else None
+        elif value is None and self.replace_nulls_with_empty:
```
Comment (ddicocco) [inline bigquery_exporter/base.py:264]: i do think it makes sense to update the name of `replace_nulls_with_empty` to `replace_nulls_with_defaults` since we're now dealing with much more than just empty strs

File: tests/test_base.py [hunk 3]
```diff
@@ -317,3 +317,157 @@ def test_table_has_data_converts_timezone_aware_datetime(self, test_exporter):
...
+
+        # Create a mock exporter with replace_nulls_with_empty=True
+        mock_client = mocker.MagicMock()
+        mock_client.get_table.return_value = mock_table
+
+        class TestExporter(BigQueryExporter):
+            model = mocker.MagicMock()
+            table_name = 'test_table'
+            fields = []
+            replace_nulls_with_empty = True
```
Comment (ddicocco) [inline tests/test_base.py]: could fixture factory this with an arg for `replace_nulls_with_empty` to cut down `TextExporter` definition redundancy

File: tests/test_base.py [hunk 4]
```diff
@@ -317,3 +317,157 @@ def test_table_has_data_converts_timezone_aware_datetime(self, test_exporter):
...
+        integer_field.name = 'integer_field'
+        integer_field.field_type = 'INTEGER'
+
+        float_field = mocker.MagicMock()
+        float_field.name = 'float_field'
+        float_field.field_type = 'FLOAT'
+
+        boolean_field = mocker.MagicMock()
+        boolean_field.name = 'boolean_field'
+        boolean_field.field_type = 'BOOLEAN'
```
Comment (ddicocco) [inline tests/test_base.py]: could add like a mock_field fixture that takes args for name & field type, again just to cut down on repetition

File: bigquery_exporter/base.py (hunk 1 above)
Comment (DevDaveFrame) [reply bigquery_exporter/base.py]: oof no yeah absolutely, I have no excuse for this one.

File: bigquery_exporter/base.py (hunk 2 above)
Comment (DevDaveFrame) [reply bigquery_exporter/base.py:264]: I'm really conflicted on this one b/c it's a good point but it would also be a breaking change that would require renaming the property on all the subclasses we've already created.

File: tests/test_base.py (hunk 3 above)
Comment (DevDaveFrame) [reply tests/test_base.py]: for sure, will do!

File: tests/test_base.py (hunk 4 above)
Comment (DevDaveFrame) [reply tests/test_base.py]: also done!

File: bigquery_exporter/clients/google_client.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            credentials: The path to the service account credentials file.
+
+        Raises:
+            BigQueryExporterInitError: If an error occurs while initializing the client.
+        """
+        try:
+            if credentials:
+                service_account_info = service_account.Credentials.from_service_account_file(credentials)
+                self.client = bigquery.Client(project=project, credentials=service_account_info)
+            else:
```
Comment (ddicocco) [inline bigquery_exporter/clients/google_client.py]: imo worth keeping the "otherwise, use default credentials" comment for this else statement– don't think it's readily apparent to anyone unfamiliar that that's what's happening

File: tests/test_base.py [hunk 6]
```diff
@@ -109,23 +89,20 @@ def test_field(self, obj):
...
-            fields = ['field_value', 'custom_field']
+        # Add the custom field method to the exporter
+        @custom_field
+        def custom_field_method(self, obj):
+            return obj.field_value * 2
 
-            @custom_field
-            def custom_field(self, obj):
-                return obj.field_value * 2
+        exporter.custom_field = custom_field_method.__get__(exporter, type(exporter))
```
Comment (ddicocco) [inline tests/test_base.py]: funky biz going on here with `self`

Comment (DevDaveFrame) [reply tests/test_base.py]: This was a misguided attempt to assign the field to the exporter instance produced by the factory. Just went back to declaring a new class from scratch.

File: bigquery_exporter/clients/google_client.py (hunk 5 above)
Comment (DevDaveFrame) [reply bigquery_exporter/clients/google_client.py]: For sure, just added back
